System/File/OsPath.hs

```diff
@@ -22,41 +44,44 @@ import qualified Data.ByteString.Lazy as BSL
 -- On POSIX systems, 'openBinaryFile' is an /interruptible operation/ as
 -- described in "Control.Exception".
 openBinaryFile :: OsPath -> IOMode -> IO Handle
-openBinaryFile fp iomode = do
-  h <- openFile fp iomode
+openBinaryFile osfp iomode = augmentError "openBinaryFile" osfp $ openBinaryFile' osfp iomode
+
+openBinaryFile' :: OsPath -> IOMode -> IO Handle
+openBinaryFile' (OsString fp) iomode =do
+  h <- P.openFile fp iomode
   hSetBinaryMode h True
   pure h
 
 -- | Run an action on a file.
 --
 -- The 'Handle' is automatically closed afther the action.
 withFile :: OsPath -> IOMode -> (Handle -> IO r) -> IO r
-withFile fp iomode action = bracket
-  (openFile fp iomode)
+withFile osfp@(OsString fp) iomode action = either ioError pure =<< (augmentError "withFile" osfp $ bracket
+  (P.openFile fp iomode)
   hClose
-  action
+  (try . action))
```

Comment:
This is the same as https://github.com/haskell/core-libraries-committee/issues/237